docs(validation): snippets split, improved, and lint

leandrodamascena opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
**Issue number:** #1370 

## Summary

### Changes

Changes snippet fictitious names to real code snippet filenames for easier discovery and collaboration to improve/correct document snippets.

Some examples have also been refactored so that the codes are complete and there are no lint/pre-commit hooks issues.

### User experience

Before:
![image](https://user-images.githubusercontent.com/4295173/184265667-d12a6110-1546-48a6-89f8-8ba381cd2134.png)
![image](https://user-images.githubusercontent.com/4295173/184265687-23ccc969-6d4f-439c-a366-186fcb27a23a.png)

After:
![image](https://user-images.githubusercontent.com/4295173/184265721-e6b36528-0c6f-44b2-913c-1f8d33ec87ea.png)
![image](https://user-images.githubusercontent.com/4295173/184265749-fce8e16b-21d1-4269-b179-3d702a9f8949.png)

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.


-----
[View rendered docs/utilities/jmespath_functions.md](https://github.com/leandrodamascena/aws-lambda-powertools-python/blob/chore/validation-docs/docs/utilities/jmespath_functions.md)
[View rendered docs/utilities/validation.md](https://github.com/leandrodamascena/aws-lambda-powertools-python/blob/chore/validation-docs/docs/utilities/validation.md)

Commit:
chore(ci): update changelog with latest changes

peterschutt committed:
docs(apigateway): removes duplicate admonition (#1426)

Commit:
chore(ci): update changelog with latest changes

peterschutt committed:
docs(parser): minor grammar fix (#1427)

Commit:
chore(ci): update changelog with latest changes

Commit:
chore(ci): update changelog with latest changes

leandrodamascena committed:
docs(validation): Refactoring examples - validator decorator

leandrodamascena committed:
docs(validation): Refactoring examples - validator decorator

leandrodamascena committed:
docs(validation): Refactoring examples - validator standalone

leandrodamascena committed:
docs(validation): Refactoring examples - validator standalone

leandrodamascena committed:
docs(jmespath_utils): Highlights

leandrodamascena committed:
docs(validation): Refactoring examples - validator standalone

leandrodamascena committed:
docs(validation): Refactoring examples - validator unwrap data

leandrodamascena committed:
docs(validation): Refactoring examples - validator unwrap data

leandrodamascena committed:
docs(validation): Refactoring examples - validator unwrap data - payload

leandrodamascena committed:
docs(validation): Refactoring examples - validator unwrap popular events

leandrodamascena committed:
fix(jmespath_util): snappy as dev dep and typing example (#1446)

Commit:
chore(ci): update changelog with latest changes

leandrodamascena committed:
docs(validation): Refactoring examples - fix classes

leandrodamascena committed:
Merge branch 'develop' into chore/validation-docs

leandrodamascena committed:
docs(validation): Refactoring examples - custom format

leandrodamascena committed:
docs(validation): Refactoring examples - custom format

heitorlessa commented:
@leandrodamascena could you rebase and fix the conflicts?

leandrodamascena commented:
> @leandrodamascena could you rebase and fix the conflicts?

Yep! Gimme some minutes and I do this.

leandrodamascena committed:
docs(validation): Rebase

leandrodamascena committed:
docs(validation): Fixing typing

heitorlessa committed:
chore: make examples tip consistent with other docs

Signed-off-by: heitorlessa <lessa@amazon.co.uk>

heitorlessa commented on examples/validation/src/getting_started_validator_decorator_function.py:
```diff
@@ -0,0 +1,28 @@
+
+    user_details: dict = {}
+
+    if event.get("username") == "lambda" and event.get("password") == "powertools":
+        user_details = User(ip=event.get("ip"), permissions=["read", "write"]).__dict__
+
+    # the body must be a object because must match OUTPUT schema, otherwise it fails
+    return {"body": user_details or None, "statusCode": 200 if user_details else 204}
```
Consider revisiting username/password idea, maybe get users instead? It's a highly visited doc and we could accidentally give the wrong impression that this is a good security mechanism.

leandrodamascena replied:
Nice! I'll refactor this here and in the other one example.

heitorlessa committed:
fix: correct eventbridge payload

heitorlessa committed:
fix: name typo and highlight

heitorlessa committed:
chore: remove EventBridge built-in envelope from earlier example

heitorlessa committed:
fix: old typo

heitorlessa committed:
chore: improve old wording

heitorlessa committed:
fix: event data access, highlight

heitorlessa committed:
fix: snippet name and highlight

heitorlessa committed:
fix: highlight SchemaValidationError

heitorlessa commented:
Looks great - thank you!! Made some quick fixes to speed up for merging. The only change pending is to rethink the `username/password` sample.

One thing you'll notice in the quick changes is the highlight. We need to be mindful of people's eyesight with an accidental excessive highlight - only highlight the absolute parts you're trying to show them about X example. For example, highlight the `SchemaValidationError` not all other exceptions.

heitorlessa reviewed (changes requested):
rethink username/password snippet

leandrodamascena committed:
docs(validation): Refactoring insecure example

leandrodamascena committed:
docs(validation): Refactoring validate example

leandrodamascena commented:
> rethink username/password snippet

I refactored this piece of code and added a integration with the Parameter utility to get data from there and raise the bar in this example. Please check if it's ok now.

Thanks for advising me on the highlights mindset. It really makes a lot of sense not to add too many highlights and make the user journey hard.

rubenfonseca commented on docs/utilities/validation.md:
````diff
@@ -31,31 +36,22 @@ We support any JSONSchema draft supported by [fastjsonschema](https://horejsek.g
-    def handler(event, context):
-        return event
-    ```
+	```python hl_lines="7 8 10 12 17 19 24 26 28 41 42 44 46 51 53"
````
I'm also a little concerned about the number of highlights when I was checking the rendered documentation. Can we switch to an example or two, maybe the most important parts?

leandrodamascena replied:
I'm not sure if it's a problem to highlight all these lines in the schema file. The reason for this is to make sure it's very clear to someone who wants to validate input and output data. The types must match 100%, otherwise it will fail.

For example: I highlighted line 24 because if the user doesn't enter a valid ipv4 address, it fails. Same for line 53, which expects an object.

Might we remove highlighting for examples and required fields (lines 7, 8, 41, 42)?

rubenfonseca commented on docs/utilities/validation.md:
```diff
@@ -67,71 +63,48 @@ It will fail fast with `SchemaValidationError` exception if event or response do
-=== "unwrapping_events.py"
+=== "getting_started_validator_unwrapping_function.py"
 
-    We use the `envelope` parameter to extract the payload inside the `detail` key before validating.
```
This comment was relevant to me as I was reading the documentation. Do you think we could include it as a code comment on the sample?

rubenfonseca commented on poetry.lock:
```diff
@@ -633,8 +633,8 @@ packaging = "*"
-test = ["flake8 (>=3.0)", "coverage"]
-dev = ["pypandoc (>=1.4)", "flake8 (>=3.0)", "coverage"]
+dev = ["coverage", "flake8 (>=3.0)", "pypandoc (>=1.4)"]
+test = ["coverage", "flake8 (>=3.0)"]
```
Funny thing that happened here, but I guess it looks better in alphabetical order :)

rubenfonseca reviewed (commented):
Hi @leandrodamascena this looks great to me. The only concern I share is too much highlighting when browsing the docs, but I guess you already working on that. Can you ping me again when you finish the changes?

leandrodamascena committed:
docs(validation): small things

leandrodamascena committed:
docs(validation): small things

rubenfonseca reviewed (approved):
Lovely job! Thank you!

rubenfonseca commented:
Not sure why CI/CD didn't run here? I would love to have mypy check the samples for errors before merging

leandrodamascena commented:
> Not sure why CI/CD didn't run here? I would love to have mypy check the samples for errors before merging

Well.. mypy and other checks are working fine locally
![image](https://user-images.githubusercontent.com/4295173/185422918-0b29390d-20ff-4782-b794-901c9ea77200.png)

rubenfonseca committed:
fix(docs): typo
